File: backups/2025-07-31_17-51-24_production_deployment_complete/routes/feature_access.py
"""
Feature Access Routes

This module provides routes for:
- Feature access checking and management
- Trial offer activation
- Subscription upgrade flows
- Educational content delivery
- Graceful feature degradation
"""

import logging
from flask import Blueprint, request, jsonify, session, current_app
from functools import wraps
from typing import Dict, Any
from backend.middleware.auth import require_auth
from backend.middleware.feature_access_middleware import require_feature_access, require_subscription_tier
from backend.services.enhanced_feature_access_service import EnhancedFeatureAccessService

logger = logging.getLogger(__name__)

# Create blueprint
feature_access_bp = Blueprint('feature_access', __name__, url_prefix='/api/features')

def handle_api_errors(f):
    """Decorator to handle API errors consistently"""
    @wraps(f)
    def decorated_function(*args, **kwargs):
        try:
            return f(*args, **kwargs)
        except Exception as e:
            logger.error(f"Error in {f.__name__}: {e}")
            return jsonify({
                'success': False,
                'error': 'internal_error',
                'message': 'Internal server error'
            }), 500
    return decorated_function

@feature_access_bp.route('/check-access/<feature_id>', methods=['GET'])
@require_auth
@handle_api_errors
def check_feature_access(feature_id: str):
    """Check if user has access to a specific feature"""
    try:
        user_id = session.get('user_id')
        if not user_id:
            return jsonify({
                'success': False,
                'error': 'authentication_required',
                'message': 'Authentication required'
            }), 401
        
        # Get feature access service
        feature_service = current_app.feature_access_service
        if not feature_service:
            return jsonify({
                'success': False,
                'error': 'service_unavailable',
                'message': 'Feature access service unavailable'
            }), 503
        
        # Check feature access
        access_result = feature_service.check_feature_access(
            user_id=str(user_id),
            feature_id=feature_id,
            context={'endpoint': request.endpoint, 'method': request.method}
        )
        
        # Return comprehensive access result
        return jsonify({
            'success': True,
            'feature_id': feature_id,
            'has_access': access_result.has_access,
            'reason': access_result.reason,
            'current_tier': access_result.current_tier,
            'required_tier': access_result.required_tier,
            'current_usage': access_result.current_usage,
            'usage_limits': access_result.usage_limits,
            'upgrade_required': access_result.upgrade_required,
            'trial_available': access_result.trial_available,
            'educational_content': access_result.educational_content,
            'alternative_suggestions': access_result.alternative_suggestions,
            'upgrade_benefits': access_result.upgrade_benefits,
            'trial_offer': access_result.trial_offer,
            'grace_period_remaining': access_result.grace_period_remaining
        })
        
    except Exception as e:
        logger.error(f"Error checking feature access: {e}")
        return jsonify({
            'success': False,
            'error': 'internal_error',
            'message': 'Internal server error'
        }), 500

@feature_access_bp.route('/summary', methods=['GET'])
@require_auth
@handle_api_errors
def get_feature_summary():
    """Get comprehensive feature summary for user"""
    try:
        user_id = session.get('user_id')
        if not user_id:
            return jsonify({
                'success': False,
                'error': 'authentication_required',
                'message': 'Authentication required'
            }), 401
        
        # Get feature access service
        feature_service = current_app.feature_access_service
        if not feature_service:
            return jsonify({
                'success': False,
                'error': 'service_unavailable',
                'message': 'Feature access service unavailable'
            }), 503
        
        # Get feature summary
        summary = feature_service.get_user_feature_summary(str(user_id))
        
        return jsonify(summary)
        
    except Exception as e:
        logger.error(f"Error getting feature summary: {e}")
        return jsonify({
            'success': False,
            'error': 'internal_error',
            'message': 'Internal server error'
        }), 500

@feature_access_bp.route('/trial/start/<feature_id>', methods=['POST'])
@require_auth
@handle_api_errors
def start_feature_trial(feature_id: str):
    """Start a trial for a premium feature"""
    try:
        user_id = session.get('user_id')
        if not user_id:
            return jsonify({
                'success': False,
                'error': 'authentication_required',
                'message': 'Authentication required'
            }), 401
        
        # Get feature access service
        feature_service = current_app.feature_access_service
        if not feature_service:
            return jsonify({
                'success': False,
                'error': 'service_unavailable',
                'message': 'Feature access service unavailable'
            }), 503
        
        # Start trial
        result = feature_service.start_feature_trial(str(user_id), feature_id)
        
        return jsonify(result)
        
    except Exception as e:
        logger.error(f"Error starting feature trial: {e}")
        return jsonify({
            'success': False,
            'error': 'internal_error',
            'message': 'Internal server error'
        }), 500

@feature_access_bp.route('/trial/status/<feature_id>', methods=['GET'])
@require_auth
@handle_api_errors
def get_trial_status(feature_id: str):
    """Get trial status for a feature"""
    try:
        user_id = session.get('user_id')
        if not user_id:
            return jsonify({
                'success': False,
                'error': 'authentication_required',
                'message': 'Authentication required'
            }), 401
        
        # Get feature access service
        feature_service = current_app.feature_access_service
        if not feature_service:
            return jsonify({
                'success': False,
                'error': 'service_unavailable',
                'message': 'Feature access service unavailable'
            }), 503
        
        # Check trial status
        trial_status = feature_service._check_trial_status(str(user_id), feature_id)
        
        return jsonify({
            'success': True,
            'feature_id': feature_id,
            'has_used_trial': trial_status['has_used_trial'],
            'trial_start': trial_status['trial_start'],
            'trial_end': trial_status['trial_end']
        })
        
    except Exception as e:
        logger.error(f"Error getting trial status: {e}")
        return jsonify({
            'success': False,
            'error': 'internal_error',
            'message': 'Internal server error'
        }), 500

@feature_access_bp.route('/upgrade/pricing', methods=['GET'])
@require_auth
@handle_api_errors
def get_upgrade_pricing():
    """Get upgrade pricing information"""
    try:
        user_id = session.get('user_id')
        if not user_id:
            return jsonify({
                'success': False,
                'error': 'authentication_required',
                'message': 'Authentication required'
            }), 401
        
        # Get feature access service
        feature_service = current_app.feature_access_service
        if not feature_service:
            return jsonify({
                'success': False,
                'error': 'service_unavailable',
                'message': 'Feature access service unavailable'
            }), 503
        
        # Get user subscription
        subscription = feature_service._get_user_subscription(str(user_id))
        if not subscription:
            return jsonify({
                'success': False,
                'error': 'no_subscription',
                'message': 'No active subscription found'
            }), 404
        
        current_tier = subscription.get('tier', 'budget')
        tier_config = feature_service.tier_configs.get(current_tier, {})
        
        return jsonify({
            'success': True,
            'current_tier': current_tier,
            'current_tier_name': tier_config.get('name', 'Unknown'),
            'current_price': tier_config.get('price', 0),
            'upgrade_path': tier_config.get('upgrade_path'),
            'upgrade_price': tier_config.get('upgrade_price'),
            'all_tiers': feature_service.tier_configs
        })
        
    except Exception as e:
        logger.error(f"Error getting upgrade pricing: {e}")
        return jsonify({
            'success': False,
            'error': 'internal_error',
            'message': 'Internal server error'
        }), 500

@feature_access_bp.route('/educational-content/<category>', methods=['GET'])
@require_auth
@handle_api_errors
def get_educational_content(category: str):
    """Get educational content for a feature category"""
    try:
        # Get feature access service
        feature_service = current_app.feature_access_service
        if not feature_service:
            return jsonify({
                'success': False,
                'error': 'service_unavailable',
                'message': 'Feature access service unavailable'
            }), 503
        
        # Get educational content
        content = feature_service._get_educational_content(category)
        
        return jsonify({
            'success': True,
            'category': category,
            'content': content
        })
        
    except Exception as e:
        logger.error(f"Error getting educational content: {e}")
        return jsonify({
            'success': False,
            'error': 'internal_error',
            'message': 'Internal server error'
        }), 500

@feature_access_bp.route('/usage/<feature_id>', methods=['GET'])
@require_auth
@handle_api_errors
def get_feature_usage(feature_id: str):
    """Get current usage for a feature"""
    try:
        user_id = session.get('user_id')
        if not user_id:
            return jsonify({
                'success': False,
                'error': 'authentication_required',
                'message': 'Authentication required'
            }), 401
        
        # Get feature access service
        feature_service = current_app.feature_access_service
        if not feature_service:
            return jsonify({
                'success': False,
                'error': 'service_unavailable',
                'message': 'Feature access service unavailable'
            }), 503
        
        # Get user subscription
        subscription = feature_service._get_user_subscription(str(user_id))
        if not subscription:
            return jsonify({
                'success': False,
                'error': 'no_subscription',
                'message': 'No active subscription found'
            }), 404
        
        current_tier = subscription.get('tier', 'budget')
        
        # Get usage information
        usage_check = feature_service._check_usage_limits(str(user_id), feature_id, current_tier)
        
        return jsonify({
            'success': True,
            'feature_id': feature_id,
            'current_tier': current_tier,
            'current_usage': usage_check['current_usage'],
            'usage_limits': usage_check['usage_limits'],
            'within_limits': usage_check['within_limits'],
            'exceeded_limits': usage_check['exceeded_limits']
        })
        
    except Exception as e:
        logger.error(f"Error getting feature usage: {e}")
        return jsonify({
            'success': False,
            'error': 'internal_error',
            'message': 'Internal server error'
        }), 500

@feature_access_bp.route('/graceful-degradation/<feature_id>', methods=['GET'])
@require_auth
@handle_api_errors
def get_graceful_degradation_info(feature_id: str):
    """Get graceful degradation information for a feature"""
    try:
        user_id = session.get('user_id')
        if not user_id:
            return jsonify({
                'success': False,
                'error': 'authentication_required',
                'message': 'Authentication required'
            }), 401
        
        # Get feature access service
        feature_service = current_app.feature_access_service
        if not feature_service:
            return jsonify({
                'success': False,
                'error': 'service_unavailable',
                'message': 'Feature access service unavailable'
            }), 503
        
        # Check feature access
        access_result = feature_service.check_feature_access(
            user_id=str(user_id),
            feature_id=feature_id,
            context={'endpoint': request.endpoint, 'method': request.method}
        )
        
        # Create graceful degradation response
        degradation_info = {
            'success': True,
            'feature_id': feature_id,
            'has_access': access_result.has_access,
            'reason': access_result.reason,
            'current_tier': access_result.current_tier,
            'required_tier': access_result.required_tier,
            'upgrade_required': access_result.upgrade_required,
            'trial_available': access_result.trial_available,
            'educational_content': access_result.educational_content,
            'alternative_suggestions': access_result.alternative_suggestions,
            'upgrade_benefits': access_result.upgrade_benefits,
            'trial_offer': access_result.trial_offer,
            'grace_period_remaining': access_result.grace_period_remaining,
            'degradation_message': _create_degradation_message(access_result),
            'alternative_features': _get_alternative_features(feature_id, access_result.current_tier),
            'upgrade_path': _get_upgrade_path(access_result.current_tier)
        }
        
        return jsonify(degradation_info)
        
    except Exception as e:
        logger.error(f"Error getting graceful degradation info: {e}")
        return jsonify({
            'success': False,
            'error': 'internal_error',
            'message': 'Internal server error'
        }), 500

def _create_degradation_message(access_result) -> str:
    """Create a user-friendly degradation message"""
    if access_result.reason == 'no_subscription':
        return "This feature requires a subscription. Start your journey with MINGUS today!"
    
    elif access_result.reason == 'upgrade_required':
        return f"This feature is available with {access_result.required_tier} tier or higher. Upgrade to unlock premium features!"
    
    elif access_result.reason == 'usage_limit_exceeded':
        return "You've reached your monthly limit for this feature. Upgrade for unlimited access!"
    
    else:
        return "This feature is not available with your current plan. Consider upgrading for more features!"

def _get_alternative_features(feature_id: str, current_tier: str) -> list:
    """Get alternative features available to the user"""
    alternatives = {
        'health_analytics': ['health_checkin', 'basic_health_tracking'],
        'ai_insights': ['basic_calculators', 'educational_content'],
        'custom_reports': ['standard_reports', 'report_templates'],
        'data_export': ['data_backup', 'manual_export'],
        'api_access': ['web_interface', 'integration_templates']
    }
    
    return alternatives.get(feature_id, [])

def _get_upgrade_path(current_tier: str) -> dict:
    """Get upgrade path information"""
    upgrade_paths = {
        'budget': {
            'next_tier': 'mid_tier',
            'price': 35.00,
            'benefits': [
                'Unlimited health check-ins',
                'Advanced AI insights',
                'Career risk management',
                'Priority support'
            ]
        },
        'mid_tier': {
            'next_tier': 'professional',
            'price': 75.00,
            'benefits': [
                'Unlimited access to all features',
                'API access',
                'Dedicated account manager',
                'Team management'
            ]
        }
    }
    
    return upgrade_paths.get(current_tier, {})

# Example protected routes with feature access control

@feature_access_bp.route('/health/checkin', methods=['POST'])
@require_auth
@require_feature_access('health_checkin')
@handle_api_errors
def submit_health_checkin():
    """Submit health check-in (requires health_checkin feature access)"""
    try:
        # This route is protected by the require_feature_access decorator
        # If access is denied, the decorator will return an appropriate response
        
        data = request.get_json()
        user_id = session.get('user_id')
        
        # Process health check-in
        # ... implementation here ...
        
        return jsonify({
            'success': True,
            'message': 'Health check-in submitted successfully'
        })
        
    except Exception as e:
        logger.error(f"Error submitting health check-in: {e}")
        return jsonify({
            'success': False,
            'error': 'internal_error',
            'message': 'Internal server error'
        }), 500

@feature_access_bp.route('/ai/insights', methods=['POST'])
@require_auth
@require_feature_access('ai_insights')
@handle_api_errors
def generate_ai_insights():
    """Generate AI insights (requires ai_insights feature access)"""
    try:
        # This route is protected by the require_feature_access decorator
        # If access is denied, the decorator will return an appropriate response
        
        data = request.get_json()
        user_id = session.get('user_id')
        
        # Generate AI insights
        # ... implementation here ...
        
        return jsonify({
            'success': True,
            'message': 'AI insights generated successfully'
        })
        
    except Exception as e:
        logger.error(f"Error generating AI insights: {e}")
        return jsonify({
            'success': False,
            'error': 'internal_error',
            'message': 'Internal server error'
        }), 500

@feature_access_bp.route('/financial/reports', methods=['POST'])
@require_auth
@require_feature_access('financial_reports')
@handle_api_errors
def generate_financial_report():
    """Generate financial report (requires financial_reports feature access)"""
    try:
        # This route is protected by the require_feature_access decorator
        # If access is denied, the decorator will return an appropriate response
        
        data = request.get_json()
        user_id = session.get('user_id')
        
        # Generate financial report
        # ... implementation here ...
        
        return jsonify({
            'success': True,
            'message': 'Financial report generated successfully'
        })
        
    except Exception as e:
        logger.error(f"Error generating financial report: {e}")
        return jsonify({
            'success': False,
            'error': 'internal_error',
            'message': 'Internal server error'
        }), 500

@feature_access_bp.route('/career/risk-assessment', methods=['POST'])
@require_auth
@require_subscription_tier('mid_tier')
@handle_api_errors
def perform_career_risk_assessment():
    """Perform career risk assessment (requires mid_tier or higher)"""
    try:
        # This route is protected by the require_subscription_tier decorator
        # If tier is insufficient, the decorator will return an appropriate response
        
        data = request.get_json()
        user_id = session.get('user_id')
        
        # Perform career risk assessment
        # ... implementation here ...
        
        return jsonify({
            'success': True,
            'message': 'Career risk assessment completed successfully'
        })
        
    except Exception as e:
        logger.error(f"Error performing career risk assessment: {e}")
        return jsonify({
            'success': False,
            'error': 'internal_error',
            'message': 'Internal server error'
        }), 500 

File: static/js/feature-access.js
/**
 * Feature Access Control - Frontend Module
 * 
 * This module provides frontend functionality for:
 * - Feature access checking and graceful degradation
 * - Upgrade prompts and trial offers
 * - Educational content display
 * - Alternative suggestions for budget users
 */

class FeatureAccessManager {
    constructor() {
        this.baseUrl = '/api/features';
        this.currentUser = null;
        this.featureCache = new Map();
        this.upgradePrompts = new Map();
        this.init();
    }

    async init() {
        try {
            // Get current user info
            await this.getCurrentUser();
            
            // Initialize upgrade prompts
            this.initializeUpgradePrompts();
            
            // Set up global error handlers
            this.setupErrorHandlers();
            
            console.log('Feature Access Manager initialized');
        } catch (error) {
            console.error('Error initializing Feature Access Manager:', error);
        }
    }

    async getCurrentUser() {
        try {
            const response = await fetch('/api/user/current', {
                method: 'GET',
                headers: {
                    'Content-Type': 'application/json'
                }
            });

            if (response.ok) {
                this.currentUser = await response.json();
            }
        } catch (error) {
            console.error('Error getting current user:', error);
        }
    }

    initializeUpgradePrompts() {
        // Health & Wellness Upgrade Prompt
        this.upgradePrompts.set('health_analytics', {
            title: 'Unlock Advanced Health Analytics',
            message: 'Get unlimited health check-ins and personalized insights to optimize your wellness journey.',
            benefits: [
                'Unlimited health check-ins',
                'Advanced health analytics',
                'Personalized wellness recommendations',
                'Health-finance correlation insights'
            ],
            cta: 'Upgrade to Mid-Tier',
            price: '$35/month'
        });

        // AI Insights Upgrade Prompt
        this.upgradePrompts.set('ai_insights', {
            title: 'Experience AI-Powered Insights',
            message: 'Get personalized recommendations and predictive analytics to make smarter financial decisions.',
            benefits: [
                '50 AI insights per month',
                'Personalized recommendations',
                'Predictive analytics',
                'Pattern recognition'
            ],
            cta: 'Upgrade to Mid-Tier',
            price: '$35/month'
        });

        // Financial Reports Upgrade Prompt
        this.upgradePrompts.set('financial_reports', {
            title: 'Unlimited Financial Reports',
            message: 'Generate as many financial reports as you need with advanced customization options.',
            benefits: [
                'Unlimited financial reports',
                'Custom report templates',
                'Advanced financial analytics',
                'White-label reports'
            ],
            cta: 'Upgrade to Mid-Tier',
            price: '$35/month'
        });

        // Career Risk Management Upgrade Prompt
        this.upgradePrompts.set('career_risk_management', {
            title: 'Protect Your Career & Income',
            message: 'Assess and manage career-related financial risks with advanced modeling and strategies.',
            benefits: [
                'Advanced risk modeling',
                'Personalized career strategies',
                'Industry trend analysis',
                'Transition planning tools'
            ],
            cta: 'Upgrade to Mid-Tier',
            price: '$35/month'
        });

        // API Access Upgrade Prompt
        this.upgradePrompts.set('api_access', {
            title: 'Unlock API Access',
            message: 'Integrate MINGUS with your own applications and workflows.',
            benefits: [
                '10,000 API requests per month',
                'Priority API support',
                'Custom integrations',
                'Webhook support'
            ],
            cta: 'Upgrade to Professional',
            price: '$75/month'
        });
    }

    setupErrorHandlers() {
        // Global error handler for feature access errors
        window.addEventListener('feature-access-error', (event) => {
            this.handleFeatureAccessError(event.detail);
        });

        // Handle 402 Payment Required responses
        window.addEventListener('fetch-error', (event) => {
            if (event.detail.status === 402) {
                this.handlePaymentRequired(event.detail);
            }
        });
    }

    async checkFeatureAccess(featureId, context = {}) {
        try {
            const cacheKey = `${featureId}_${JSON.stringify(context)}`;
            
            // Check cache first
            if (this.featureCache.has(cacheKey)) {
                const cached = this.featureCache.get(cacheKey);
                if (Date.now() - cached.timestamp < 60000) { // 1 minute cache
                    return cached.data;
                }
            }

            const response = await fetch(`${this.baseUrl}/check-access/${featureId}`, {
                method: 'GET',
                headers: {
                    'Content-Type': 'application/json'
                }
            });

            const result = await response.json();

            // Cache the result
            this.featureCache.set(cacheKey, {
                data: result,
                timestamp: Date.now()
            });

            return result;
        } catch (error) {
            console.error('Error checking feature access:', error);
            return {
                success: false,
                error: 'network_error',
                message: 'Unable to check feature access'
            };
        }
    }

    async getFeatureSummary() {
        try {
            const response = await fetch(`${this.baseUrl}/summary`, {
                method: 'GET',
                headers: {
                    'Content-Type': 'application/json'
                }
            });

            return await response.json();
        } catch (error) {
            console.error('Error getting feature summary:', error);
            return {
                success: false,
                error: 'network_error',
                message: 'Unable to get feature summary'
            };
        }
    }

    async startFeatureTrial(featureId) {
        try {
            const response = await fetch(`${this.baseUrl}/trial/start/${featureId}`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                }
            });

            const result = await response.json();

            if (result.success) {
                this.showTrialStartedModal(featureId, result);
            }

            return result;
        } catch (error) {
            console.error('Error starting feature trial:', error);
            return {
                success: false,
                error: 'network_error',
                message: 'Unable to start trial'
            };
        }
    }

    async getGracefulDegradationInfo(featureId) {
        try {
            const response = await fetch(`${this.baseUrl}/graceful-degradation/${featureId}`, {
                method: 'GET',
                headers: {
                    'Content-Type': 'application/json'
                }
            });

            return await response.json();
        } catch (error) {
            console.error('Error getting graceful degradation info:', error);
            return {
                success: false,
                error: 'network_error',
                message: 'Unable to get degradation info'
            };
        }
    }

    handleFeatureAccessError(error) {
        const { featureId, reason, currentTier, requiredTier } = error;

        switch (reason) {
            case 'subscription_required':
                this.showSubscriptionRequiredModal(featureId);
                break;
            case 'upgrade_required':
                this.showUpgradeRequiredModal(featureId, currentTier, requiredTier);
                break;
            case 'usage_limit_exceeded':
                this.showUsageLimitModal(featureId, error.currentUsage, error.usageLimits);
                break;
            default:
                this.showGenericAccessDeniedModal(featureId);
        }
    }

    handlePaymentRequired(error) {
        const featureId = this.extractFeatureIdFromUrl();
        if (featureId) {
            this.showSubscriptionRequiredModal(featureId);
        }
    }

    showSubscriptionRequiredModal(featureId) {
        const modal = this.createModal({
            title: 'Subscription Required',
            message: 'A subscription is required to access this feature. Start your journey with MINGUS today!',
            type: 'subscription_required',
            featureId: featureId,
            actions: [
                {
                    text: 'View Plans',
                    action: () => this.navigateToPricing(),
                    primary: true
                },
                {
                    text: 'Learn More',
                    action: () => this.showEducationalContent(featureId),
                    secondary: true
                },
                {
                    text: 'Cancel',
                    action: () => this.closeModal(),
                    secondary: true
                }
            ]
        });

        this.showModal(modal);
    }

    showUpgradeRequiredModal(featureId, currentTier, requiredTier) {
        const prompt = this.upgradePrompts.get(featureId);
        if (!prompt) {
            this.showGenericUpgradeModal(featureId, currentTier, requiredTier);
            return;
        }

        const modal = this.createModal({
            title: prompt.title,
            message: prompt.message,
            type: 'upgrade_required',
            featureId: featureId,
            benefits: prompt.benefits,
            price: prompt.price,
            actions: [
                {
                    text: prompt.cta,
                    action: () => this.navigateToUpgrade(currentTier),
                    primary: true
                },
                {
                    text: 'Start Trial',
                    action: () => this.startFeatureTrial(featureId),
                    secondary: true
                },
                {
                    text: 'Learn More',
                    action: () => this.showEducationalContent(featureId),
                    secondary: true
                },
                {
                    text: 'Cancel',
                    action: () => this.closeModal(),
                    secondary: true
                }
            ]
        });

        this.showModal(modal);
    }

    showUsageLimitModal(featureId, currentUsage, usageLimits) {
        const modal = this.createModal({
            title: 'Usage Limit Reached',
            message: 'You have reached your monthly usage limit for this feature. Upgrade for unlimited access!',
            type: 'usage_limit',
            featureId: featureId,
            currentUsage: currentUsage,
            usageLimits: usageLimits,
            actions: [
                {
                    text: 'Upgrade Now',
                    action: () => this.navigateToUpgrade(),
                    primary: true
                },
                {
                    text: 'View Usage',
                    action: () => this.showUsageDetails(featureId),
                    secondary: true
                },
                {
                    text: 'Cancel',
                    action: () => this.closeModal(),
                    secondary: true
                }
            ]
        });

        this.showModal(modal);
    }

    showTrialStartedModal(featureId, trialInfo) {
        const modal = this.createModal({
            title: 'Trial Started!',
            message: `Your ${trialInfo.duration_days}-day trial for this feature has begun. Enjoy exploring the premium features!`,
            type: 'trial_started',
            featureId: featureId,
            trialInfo: trialInfo,
            actions: [
                {
                    text: 'Start Using',
                    action: () => this.closeModal(),
                    primary: true
                },
                {
                    text: 'View Trial Status',
                    action: () => this.showTrialStatus(featureId),
                    secondary: true
                }
            ]
        });

        this.showModal(modal);
    }

    showEducationalContent(featureId) {
        this.getGracefulDegradationInfo(featureId).then(result => {
            if (result.success && result.educational_content) {
                const modal = this.createModal({
                    title: 'Learn About Premium Features',
                    message: result.educational_content,
                    type: 'educational',
                    featureId: featureId,
                    alternativeSuggestions: result.alternative_suggestions,
                    actions: [
                        {
                            text: 'Upgrade Now',
                            action: () => this.navigateToUpgrade(),
                            primary: true
                        },
                        {
                            text: 'Try Alternatives',
                            action: () => this.showAlternativeFeatures(featureId, result.alternative_suggestions),
                            secondary: true
                        },
                        {
                            text: 'Close',
                            action: () => this.closeModal(),
                            secondary: true
                        }
                    ]
                });

                this.showModal(modal);
            }
        });
    }

    showAlternativeFeatures(featureId, alternatives) {
        const modal = this.createModal({
            title: 'Alternative Features',
            message: 'Here are some alternative features you can use with your current plan:',
            type: 'alternatives',
            featureId: featureId,
            alternatives: alternatives,
            actions: [
                {
                    text: 'Explore Alternatives',
                    action: () => this.navigateToAlternatives(alternatives),
                    primary: true
                },
                {
                    text: 'Upgrade Instead',
                    action: () => this.navigateToUpgrade(),
                    secondary: true
                },
                {
                    text: 'Close',
                    action: () => this.closeModal(),
                    secondary: true
                }
            ]
        });

        this.showModal(modal);
    }

    createModal(config) {
        const modal = document.createElement('div');
        modal.className = 'feature-access-modal';
        modal.innerHTML = this.generateModalHTML(config);
        
        // Add event listeners
        this.addModalEventListeners(modal, config);
        
        return modal;
    }

    generateModalHTML(config) {
        const { title, message, type, actions, benefits, price, currentUsage, usageLimits, alternatives } = config;

        let contentHTML = `<p>${message}</p>`;

        // Add benefits list
        if (benefits && benefits.length > 0) {
            contentHTML += `
                <div class="benefits-list">
                    <h4>Benefits:</h4>
                    <ul>
                        ${benefits.map(benefit => `<li>${benefit}</li>`).join('')}
                    </ul>
                </div>
            `;
        }

        // Add usage information
        if (currentUsage && usageLimits) {
            contentHTML += `
                <div class="usage-info">
                    <h4>Current Usage:</h4>
                    <p>You've used ${currentUsage[Object.keys(currentUsage)[0]]} of your monthly limit.</p>
                </div>
            `;
        }

        // Add alternatives
        if (alternatives && alternatives.length > 0) {
            contentHTML += `
                <div class="alternatives-list">
                    <h4>Alternative Features:</h4>
                    <ul>
                        ${alternatives.map(alt => `<li>${alt}</li>`).join('')}
                    </ul>
                </div>
            `;
        }

        // Add price if available
        if (price) {
            contentHTML += `<div class="price-info"><strong>${price}</strong></div>`;
        }

        const actionsHTML = actions.map(action => {
            const buttonClass = action.primary ? 'btn-primary' : 'btn-secondary';
            return `<button class="btn ${buttonClass}" data-action="${action.text.toLowerCase().replace(' ', '_')}">${action.text}</button>`;
        }).join('');

        return `
            <div class="modal-overlay">
                <div class="modal-content">
                    <div class="modal-header">
                        <h3>${title}</h3>
                        <button class="modal-close" data-action="close">&times;</button>
                    </div>
                    <div class="modal-body">
                        ${contentHTML}
                    </div>
                    <div class="modal-footer">
                        ${actionsHTML}
                    </div>
                </div>
            </div>
        `;
    }

    addModalEventListeners(modal, config) {
        const { actions } = config;

        // Close button
        modal.querySelector('.modal-close').addEventListener('click', () => {
            this.closeModal();
        });

        // Action buttons
        actions.forEach(action => {
            const button = modal.querySelector(`[data-action="${action.text.toLowerCase().replace(' ', '_')}"]`);
            if (button) {
                button.addEventListener('click', action.action);
            }
        });

        // Overlay click to close
        modal.querySelector('.modal-overlay').addEventListener('click', (e) => {
            if (e.target === e.currentTarget) {
                this.closeModal();
            }
        });
    }

    showModal(modal) {
        document.body.appendChild(modal);
        document.body.style.overflow = 'hidden';
        
        // Add animation
        setTimeout(() => {
            modal.classList.add('show');
        }, 10);
    }

    closeModal() {
        const modal = document.querySelector('.feature-access-modal');
        if (modal) {
            modal.classList.remove('show');
            setTimeout(() => {
                document.body.removeChild(modal);
                document.body.style.overflow = '';
            }, 300);
        }
    }

    navigateToPricing() {
        window.location.href = '/pricing';
    }

    navigateToUpgrade(currentTier = null) {
        const upgradeUrl = currentTier ? `/upgrade?from=${currentTier}` : '/upgrade';
        window.location.href = upgradeUrl;
    }

    navigateToAlternatives(alternatives) {
        // Navigate to a page showing alternative features
        window.location.href = '/features/alternatives';
    }

    showUsageDetails(featureId) {
        window.location.href = `/features/usage/${featureId}`;
    }

    showTrialStatus(featureId) {
        window.location.href = `/features/trial/status/${featureId}`;
    }

    extractFeatureIdFromUrl() {
        const path = window.location.pathname;
        const featurePatterns = [
            /\/health\/checkin/,
            /\/ai\/insights/,
            /\/financial\/reports/,
            /\/career\/risk/,
            /\/data\/export/
        ];

        for (const pattern of featurePatterns) {
            if (pattern.test(path)) {
                return path.split('/')[2]; // Extract feature ID
            }
        }

        return null;
    }

    // Utility method to check feature access before making API calls
    async ensureFeatureAccess(featureId, fallbackAction = null) {
        const accessResult = await this.checkFeatureAccess(featureId);
        
        if (!accessResult.success || !accessResult.has_access) {
            // Dispatch error event
            window.dispatchEvent(new CustomEvent('feature-access-error', {
                detail: accessResult
            }));
            
            // Execute fallback action if provided
            if (fallbackAction && typeof fallbackAction === 'function') {
                fallbackAction(accessResult);
            }
            
            return false;
        }
        
        return true;
    }

    // Method to gracefully degrade feature functionality
    async gracefulDegrade(featureId, fullFunction, degradedFunction) {
        const hasAccess = await this.ensureFeatureAccess(featureId);
        
        if (hasAccess) {
            return fullFunction();
        } else {
            return degradedFunction();
        }
    }
}

// Initialize the feature access manager when the page loads
document.addEventListener('DOMContentLoaded', () => {
    window.featureAccessManager = new FeatureAccessManager();
});

// Export for use in other modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = FeatureAccessManager;
} 